### ISSUE TITLE
PHtrans: something wrong with electrodes

### ISSUE THREAD

**Issue by tomato151** [zerothi/sisl#359]
Dear Nick,

I tried to calculate the dynamical matrix using siesta whose input file contains the same geometry with that in ex_03.py :

```
SystemLabel DEVICE
MD.TypeOfRun        FC
IsBulkFC              F

%block kgrid.MonkhorstPack
 5 0 0 0.
 0 1 0 0.
 0 0 1 0.
%endblock
 
# This is the remaining default SIESTA options
PAO.BasisSize         DZP
NumberOfSpecies 1
%block ChemicalSpeciesLabel
 1 6 C
%endblock ChemicalSpeciesLabel
XC.functional     GGA
XC.authors        PBE
MeshCutoff        150.000000 Ry
FilterCutoff      110.000000 Ry

ElectronicTemperature   300 K

MinSCFIterations       3
SCF.DM.Tolerance       0.0001
# Mixing parameters:
SCF.Mixer.Weight       0.10
SCF.Mixer.History      12
SCF.Mix.First          true
DM.UseSaveDM           true
MD.NumCGSteps 0

SaveHS                          T
LatticeConstant 1.0 Ang
%block LatticeVectors
 17.040000000 0.00000000 0.00000000
 0.00000000 19.67609717 0.00000000
 0.00000000 0.00000000 15.00000000
%endblock LatticeVectors

NumberOfAtoms 128
AtomicCoordinatesFormat Ang
%block AtomicCoordinatesAndAtomicSpecies
  0.00000 0.00000 0.00000       1 
  2.84000 0.00000 0.00000       1
  0.71000 1.22976 0.00000       1
  2.13000 1.22976 0.00000       1
  0.00000 2.45951 0.00000       1
  2.84000 2.45951 0.00000       1
  0.71000 3.68927 0.00000       1
  2.13000 3.68927 0.00000       1
  0.00000 4.91902 0.00000       1
  2.84000 4.91902 0.00000       1
  0.71000 6.14878 0.00000       1
  2.13000 6.14878 0.00000       1
  0.00000 7.37854 0.00000       1
  2.84000 7.37854 0.00000       1
  0.71000 8.60829 0.00000       1
  2.13000 8.60829 0.00000       1
  0.00000 9.83805 0.00000       1
  2.84000 9.83805 0.00000       1
  0.71000 11.06780 0.00000      1
  2.13000 11.06780 0.00000      1
  0.00000 12.29756 0.00000      1
  2.84000 12.29756 0.00000      1
  0.71000 13.52732 0.00000      1
  2.13000 13.52732 0.00000      1
  0.00000 14.75707 0.00000      1
  2.84000 14.75707 0.00000      1
  0.71000 15.98683 0.00000      1
  2.13000 15.98683 0.00000      1
  0.00000 17.21659 0.00000      1
  2.84000 17.21659 0.00000      1
  0.71000 18.44634 0.00000      1
  2.13000 18.44634 0.00000      1
  4.26000 0.00000 0.00000       1
  7.10000 0.00000 0.00000       1
  4.97000 1.22976 0.00000       1
  6.39000 1.22976 0.00000       1
  4.26000 2.45951 0.00000       1
  7.10000 2.45951 0.00000       1
  4.97000 3.68927 0.00000       1
  6.39000 3.68927 0.00000       1
  4.26000 4.91902 0.00000       1
  7.10000 4.91902 0.00000       1
  4.97000 6.14878 0.00000       1
  6.39000 6.14878 0.00000       1
  4.26000 7.37854 0.00000       1
  7.10000 7.37854 0.00000       1
  4.97000 8.60829 0.00000       1
  6.39000 8.60829 0.00000       1
  4.26000 9.83805 0.00000       1
  7.10000 9.83805 0.00000       1
  4.97000 11.06780 0.00000      1
  6.39000 11.06780 0.00000      1
  4.26000 12.29756 0.00000      1
  7.10000 12.29756 0.00000      1
  4.97000 13.52732 0.00000      1
  6.39000 13.52732 0.00000      1
  4.26000 14.75707 0.00000      1
  7.10000 14.75707 0.00000      1
  4.97000 15.98683 0.00000      1
  6.39000 15.98683 0.00000      1
  4.26000 17.21659 0.00000      1
  7.10000 17.21659 0.00000      1
  4.97000 18.44634 0.00000      1
  6.39000 18.44634 0.00000      1
  8.52000 0.00000 0.00000       1
  11.36000 0.00000 0.00000      1
  9.23000 1.22976 0.00000       1
  10.65000 1.22976 0.00000      1
  8.52000 2.45951 0.00000       1
  11.36000 2.45951 0.00000      1
  9.23000 3.68927 0.00000       1
  10.65000 3.68927 0.00000      1
  8.52000 4.91902 0.00000       1
  11.36000 4.91902 0.00000      1
  9.23000 6.14878 0.00000       1
  10.65000 6.14878 0.00000      1
  8.52000 7.37854 0.00000       1
  11.36000 7.37854 0.00000      1
  9.23000 8.60829 0.00000       1
  10.65000 8.60829 0.00000      1
  8.52000 9.83805 0.00000       1
  11.36000 9.83805 0.00000      1
  9.23000 11.06780 0.00000      1
  10.65000 11.06780 0.0000      1
  8.52000 12.29756 0.00000      1
  11.36000 12.29756 0.0000      1
  9.23000 13.52732 0.00000      1
  10.65000 13.52732 0.0000      1
  8.52000 14.75707 0.00000      1
  11.36000 14.75707 0.0000      1
  9.23000 15.98683 0.00000      1
  10.65000 15.98683 0.0000      1
  8.52000 17.21659 0.00000      1
  11.36000 17.21659 0.0000      1
  9.23000 18.44634 0.00000      1
  10.65000 18.44634 0.0000      1
  12.78000 0.00000 0.00000      1
  15.62000 0.00000 0.00000      1
  13.49000 1.22976 0.00000      1
  14.91000 1.22976 0.00000      1
  12.78000 2.45951 0.00000      1
  15.62000 2.45951 0.00000      1
  13.49000 3.68927 0.00000      1
  14.91000 3.68927 0.00000      1
  12.78000 4.91902 0.00000      1
  15.62000 4.91902 0.00000      1
  13.49000 6.14878 0.00000      1
  14.91000 6.14878 0.00000      1
  12.78000 7.37854 0.00000      1
  15.62000 7.37854 0.00000      1
  13.49000 8.60829 0.00000      1
  14.91000 8.60829 0.00000      1
  12.78000 9.83805 0.00000      1
  15.62000 9.83805 0.00000      1
  13.49000 11.06780 0.00000     1
  14.91000 11.06780 0.00000     1
  12.78000 12.29756 0.00000     1
  15.62000 12.29756 0.00000     1
  13.49000 13.52732 0.00000     1
  14.91000 13.52732 0.00000     1
  12.78000 14.75707 0.00000     1
  15.62000 14.75707 0.00000     1
  13.49000 15.98683 0.00000     1
  14.91000 15.98683 0.00000     1
  12.78000 17.21659 0.00000     1
  15.62000 17.21659 0.00000     1
  13.49000 18.44634 0.00000     1
  14.91000 18.44634 0.00000     1
%endblock AtomicCoordinatesAndAtomicSpecies

```

Siesta finished normally and then I used sisl to get the dynamical matrices:

```
import sisl
dyn = sisl.get_sile('RUN.fdf').read_dynamical_matrix()
dev = dyn.cut(4, 0)
dev.write('DEVICE_zz.nc')
el = dev.cut(4, 1)
el.write('ELEC_zz.nc')
```
But when I run phtrans, the following errors occurs:

Electrode information:

  <sparsity:./ELEC_zz.nc
    nrows_g=24 nrows=24 sparsity=49.0000 nnzs=28224, refcount: 3>
  <sparsity:(TM [--, 0,--] of: ./ELEC_zz.nc)
    nrows_g=24 nrows=24 sparsity=7.0000 nnzs=4032, refcount: 3>
  <sparsity:(TM [--,-1,--] of: ./ELEC_zz.nc)
    nrows_g=24 nrows=24 sparsity=7.0000 nnzs=4032, refcount: 3>
 Left principal cell is extending out with 4032 elements:
    Atom 1 connects with atom 7
    Orbital 1 connects with orbital 19
    Hamiltonian value: |H(1,859)|@R=-2 =  0.349E-05 eV
    Overlap          :  S(1,859)|@R=-2 =   0.00    
Electrode connectivity is not perfect, refer to the manual for achieving a perfect electrode.
Stopping Program from Node:    0

I would greatly appreciate your help!

**Comment by zerothi** [comment]
Hi,

1. I would be very careful about using the dynamical matrix from force constants. There are some corner cases where sisl doesn't calculate the correct dynamical matrix (missing repetition factors). I would suspect it to work for 1D systems, but it is not thoroughly tested.
2. Another thing is that DFT phonons are much longer range than empirical GULP simulations. So you should increase the system size to ensure that there are only nearest neighbour coupling in the electrodes. This is the error you see here. You have the primary cell connected to the 2nd nearest cell. Although the hopping element seems rather weak, for dynamical matrices small values tend to be rather important, so you should be very careful here.

Bottom line, I wouldn't, at this point, trust sisl dynamical matrix conversion.

**Comment by tomato151** [comment]
Got it.Thank you for your help!
